File: zeon/converter.py
import json
import os
from .parse import loads
from .stringify import dumps

class Converter:
    def __init__(self, data_or_path: str):
        self.raw_data = data_or_path
        self.is_file = False
        self.file_path = None
        
        # Check if the input is an existing file path
        if isinstance(data_or_path, str) and len(data_or_path) < 1000 and os.path.exists(data_or_path):
            self.is_file = True
            self.file_path = data_or_path
            with open(data_or_path, 'r', encoding='utf-8') as f:
                self.content = f.read()
        else:
            self.content = data_or_path

    def to_json(self, out_path: str = None, **kwargs) -> str:
        """
        Parses ZEON content and returns JSON string.
        If out_path is provided, it saves to the file.
        """
        parsed_data = loads(self.content)
        
        if not kwargs and 'indent' not in kwargs:
            result = json.dumps(parsed_data, separators=(',', ':'))
        else:
            result = json.dumps(parsed_data, **kwargs)
            
        if out_path:
            # Se for apenas um nome de arquivo (sem barra de pasta) e a origem for um arquivo,
            # salva na mesma pasta do arquivo original.
            if self.is_file and not os.path.dirname(out_path):
                original_dir = os.path.dirname(self.file_path)
                out_path = os.path.join(original_dir, out_path)
                
            with open(out_path, 'w', encoding='utf-8') as f:
                f.write(result)
                
        return result

    def to_yaml(self, out_path: str = None, **kwargs) -> str:
        """
        Parses ZEON content and returns YAML string.
        If out_path is provided, it saves to the file.
        """
        import yaml
        parsed_data = loads(self.content)
        
        if not kwargs and 'sort_keys' not in kwargs:
            result = yaml.safe_dump(parsed_data, sort_keys=False)
        else:
            result = yaml.safe_dump(parsed_data, **kwargs)
            
        if out_path:
            if self.is_file and not os.path.dirname(out_path):
                original_dir = os.path.dirname(self.file_path)
                out_path = os.path.join(original_dir, out_path)
                
            with open(out_path, 'w', encoding='utf-8') as f:
                f.write(result)
                
        return result

    def to_zeon(self, out_path: str = None, **kwargs) -> str:
        """
        Parses JSON or YAML content and returns ZEON string.
        If out_path is provided, it saves to the file.
        """
        try:
            parsed_data = json.loads(self.content)
        except json.JSONDecodeError:
            import yaml
            parsed_data = yaml.safe_load(self.content)
            
        result = dumps(parsed_data, **kwargs)
        
        if out_path:
            if self.is_file and not os.path.dirname(out_path):
                original_dir = os.path.dirname(self.file_path)
                out_path = os.path.join(original_dir, out_path)
                
            with open(out_path, 'w', encoding='utf-8') as f:
                f.write(result)
                
        return result

def convert(data_or_path: str) -> Converter:
    """
    Starts a fluent conversion process. 
    Accepts either a raw string or a file path.
    """
    return Converter(data_or_path)

File: parsers/javascript/src/converter.ts
import * as fs from 'fs';
import * as path from 'path';
import * as yaml from 'js-yaml';
import { parse } from './parse';
import { dumps } from './stringify';

export class Converter {
    private isFile: boolean = false;
    private filePath: string | null = null;
    private content: string;

    constructor(dataOrPath: string) {
        if (dataOrPath.length < 1000 && fs.existsSync(dataOrPath)) {
            this.isFile = true;
            this.filePath = path.resolve(dataOrPath);
            this.content = fs.readFileSync(this.filePath, 'utf-8');
        } else {
            this.content = dataOrPath;
        }
    }

    private _saveIfRequested(outPath: string | undefined, result: string) {
        if (outPath) {
            let finalPath = outPath;
            if (this.isFile && this.filePath && !path.dirname(outPath).includes(path.sep)) {
                const originalDir = path.dirname(this.filePath);
                finalPath = path.join(originalDir, outPath);
            }
            fs.writeFileSync(finalPath, result, 'utf-8');
        }
    }

    toJson(outPath?: string, space: number = 0): string {
        const parsedData = this._loadData();
        const result = JSON.stringify(parsedData, null, space);
        this._saveIfRequested(outPath, result);
        return result;
    }

    toYaml(outPath?: string): string {
        const parsedData = this._loadData();
        const result = yaml.dump(parsedData, { indent: 2, lineWidth: -1 });
        this._saveIfRequested(outPath, result);
        return result;
    }

    toZeon(outPath?: string): string {
        const parsedData = this._loadData();
        const result = dumps(parsedData);
        this._saveIfRequested(outPath, result);
        return result;
    }

    private _loadData(): any {
        // Se a entrada for um arquivo JSON ou YAML
        if (this.isFile && this.filePath) {
            if (this.filePath.endsWith('.json')) {
                return JSON.parse(this.content);
            } else if (this.filePath.endsWith('.yaml') || this.filePath.endsWith('.yml')) {
                return yaml.load(this.content);
            }
        }
        
        // Se for uma string pura de JSON
        const trimmed = this.content.trim();
        if ((trimmed.startsWith('{') && trimmed.endsWith('}')) || 
            (trimmed.startsWith('[') && trimmed.endsWith(']'))) {
            try {
                return JSON.parse(trimmed);
            } catch (e) {
                // fall back to ZEON if JSON parsing fails
            }
        }

        // Caso contrário, é ZEON
        return parse(this.content);
    }
}

export function convert(dataOrPath: string): Converter {
    return new Converter(dataOrPath);
}
